File: java/starter/issuer/src/main/java/network/t0/pay/issuer/internal/PaymentExpired.java
package network.t0.pay.issuer.internal;

import io.grpc.StatusRuntimeException;
import network.t0.pay.proto.tzero.v1.pay.IssuerPaymentExpiredRequest;
import network.t0.pay.proto.tzero.v1.pay.IssuerPaymentExpiredResponse;
import network.t0.pay.proto.tzero.v1.pay.IssuerServiceGrpc;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.concurrent.TimeUnit;

/**
 * §14 PaymentExpired — the reservation closed with no valid payment and you have
 * released the deposit addresses back to the pool.
 *
 * <p>Confirmation, not a trigger: t-0 expires the intent on its own clock at
 * {@code expiresAt} and tells the acquirer via §15 regardless of this call.
 *
 * <p>Idempotency key: {@code paymentIntentId}.
 */
public final class PaymentExpired {

    private static final Logger log = LoggerFactory.getLogger(PaymentExpired.class);

    private static final int TIMEOUT_SECONDS = 10;

    public static Outcome<IssuerPaymentExpiredResponse.Accepted> report(
            IssuerServiceGrpc.IssuerServiceBlockingStub t0,
            long paymentIntentId,
            Instant expiredAt) {

        IssuerPaymentExpiredRequest request = IssuerPaymentExpiredRequest.newBuilder()
                .setPaymentIntentId(paymentIntentId)
                .setExpiredAt(Times.from(expiredAt))
                .build();

        try {
            IssuerPaymentExpiredResponse response =
                    t0.withDeadlineAfter(TIMEOUT_SECONDS, TimeUnit.SECONDS).paymentExpired(request);

            switch (response.getResultCase()) {
                case ACCEPTED -> {
                    log.info("§14 accepted: intent={} released", paymentIntentId);
                    return new Outcome.Accepted<>(response.getAccepted());
                }
                case REJECTED -> {
                    // UNKNOWN_INTENT — t-0 never opened this intent. Look at where the id
                    // came from rather than resending.
                    String reason = response.getRejected().getReason().name();
                    log.warn("§14 rejected for intent {}: {}", paymentIntentId, reason);
                    return new Outcome.Rejected<>(reason);
                }
                default -> {
                    return new Outcome.Rejected<>("response carried no result variant");
                }
            }
        } catch (StatusRuntimeException e) {
            log.error("§14 failed for intent {}: {}", paymentIntentId, e.getStatus());
            return new Outcome.Unknown<>(e.getStatus().toString());
        }
    }

    private PaymentExpired() {
    }
}

File: node/starter/issuer/src/internal/payment_expired.ts
import { timestampFromDate } from "@bufbuild/protobuf/wkt";
import {
  type Client,
  type IssuerPaymentExpiredResponse_Accepted,
  IssuerPaymentExpiredResponse_Rejected_Reason,
  type IssuerService,
} from "@t-0/usdt-pay-sdk";
import { accepted, noResultVariant, outcomeFromError, rejected, type Outcome } from "./outcome.js";

const TIMEOUT_MS = 10_000;

/**
 * §14 PaymentExpired — the reservation closed with no valid payment and you have
 * released the deposit addresses back to the pool.
 *
 * Confirmation, not a trigger: t-0 expires the intent on its own clock at `expiresAt`
 * and tells the acquirer via §15 regardless of this call.
 *
 * Idempotency key: `paymentIntentId`.
 */
export async function reportPaymentExpired(
  t0: Client<typeof IssuerService>,
  expiry: { paymentIntentId: bigint; expiredAt: Date },
): Promise<Outcome<IssuerPaymentExpiredResponse_Accepted>> {
  try {
    const response = await t0.paymentExpired(
      {
        paymentIntentId: expiry.paymentIntentId,
        expiredAt: timestampFromDate(expiry.expiredAt),
      },
      { timeoutMs: TIMEOUT_MS },
    );

    switch (response.result.case) {
      case "accepted":
        console.log(`§14 accepted: intent=${expiry.paymentIntentId} released`);
        return accepted(response.result.value);

      case "rejected": {
        // UNKNOWN_INTENT — t-0 never opened this intent. Look at where the id came
        // from rather than resending.
        const { reason } = response.result.value;
        const name = IssuerPaymentExpiredResponse_Rejected_Reason[reason] ?? String(reason);
        console.warn(`§14 rejected for intent ${expiry.paymentIntentId}: ${name}`);
        return rejected(name);
      }

      default:
        return noResultVariant();
    }
  } catch (error) {
    console.error(`§14 failed for intent ${expiry.paymentIntentId}:`, error);
    return outcomeFromError(error);
  }
}
